PSScriptAnalyzer fixes by category

bergmeister opened the pull request:
This is the updated version of PR #4213, in which it was requested to split the commits on a per kind category.
The intend is to help with fixing #3791. The fixed PSScriptAnalyzer warnings are PSPossibleIncorrectComparisonWithNull and PSAvoidUsingCmdletAliases for ForEach-Object, Where-Object and Select-Object since those warning types were responsible for the majority of warnings.

bergmeister committed:
Fix PSScriptAnalyzer warnings of type PSAvoidUsingCmdletAliases for 'ForEach-Object' (alias is '%' or 'foreach')

bergmeister committed:
Fix PSScriptAnalyzer warnings of type PSAvoidUsingCmdletAliases for 'Where-Object' (alias is '?' or 'where')

bergmeister committed:
Fix PSScriptAnalyzer warnings of type PSAvoidUsingCmdletAliases for 'Select-Object' (alias is 'select')

bergmeister committed:
Fix PSScriptAnalyzer warnings of type PSPossibleIncorrectComparisonWithNull. Essentially, $null has to be on the left hand side when using it for comparison.
A Test in ParameterBinding.Tests.ps1 needed adapting as this test used to rely on the wrong null comparison
It was also suggested in the initial PR #4205 to replace a subset of tests of kind '($object -eq $null) | Should Be $true' with '($null -eq $object) | Should BeNullOrEmpty'

msftclas commented:
@bergmeister,
Thanks for having already signed the Contribution License Agreement. Your agreement was validated by Microsoft. We will now review your pull request.
_Thanks,
Microsoft Pull Request Bot_

daxian-dbw commented:
@bergmeister Thank you so much for taking extra effort splitting the original PR by the fix categories!

iSazonov commented on build.psm1:
```diff
@@ -668,7 +668,7 @@ function Get-PesterTag {
     $warnings = @()
 
-    get-childitem -Recurse $testbase -File |?{$_.name -match "tests.ps1"}| ForEach-Object {
+    get-childitem -Recurse $testbase -File |Where-Object {$_.name -match "tests.ps1"}| ForEach-Object {
```
Please add a space before Where-Object and after `"tests.ps1"}`

iSazonov commented on demos/Apache/apache-demo.ps1:
```diff
@@ -2,7 +2,7 @@ Import-Module $PSScriptRoot/Apache/Apache.psm1
 #list Apache Modules
 Write-Host -Foreground Blue "Get installed Apache Modules like *proxy* and Sort by name"
-Get-ApacheModule |Where {$_.ModuleName -like "*proxy*"}|Sort-Object ModuleName | Out-Host
+Get-ApacheModule | Where-Object {$_.ModuleName -like "*proxy*"}|Sort-Object ModuleName | Out-Host
```
Please add spaces in `"*proxy*"}|Sort-Object`.

iSazonov commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/ImplicitRemotingCommands.cs:
```diff
@@ -2139,11 +2139,11 @@ private void GenerateHelperFunctionsWriteMessage(TextWriter writer)
         $script:PSSession = $PSSession
 
-        if ($createdByModule -and ($script:PSSession -ne $null))
+        if ($null -ne $createdByModule -and ($script:PSSession))
```
Please correct.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Implicit.Remoting.Tests.ps1:
```diff
@@ -417,7 +417,7 @@ Describe "Proxy module is usable when the original runspace is no longer around"
         It "Verifies Remove-Module removed the runspace that was automatically created" -Pending {
             Remove-Module $module -Force
-            ((Get-PSSession -InstanceId $internalSession.InstanceId -ErrorAction SilentlyContinue) -eq $null) | Should Be $true
+            ($null -eq (Get-PSSession -InstanceId $internalSession.InstanceId -ErrorAction SilentlyContinue)) | Should Be $true
```
Please change to `Should BeNullOrEmpty`.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Implicit.Remoting.Tests.ps1:
```diff
@@ -452,7 +452,7 @@ Describe "Proxy module is usable when the original runspace is no longer around"
         It "Verifies Remove-Module removes automatically created runspace" -Pending {
             Remove-Module $module -Force
-            ((Get-PSSession -InstanceId $internalSession.InstanceId -ErrorAction SilentlyContinue) -eq $null) | Should Be $true
+            ($null -eq (Get-PSSession -InstanceId $internalSession.InstanceId -ErrorAction SilentlyContinue)) | Should Be $true
```
Please change to `Should BeNullOrEmpty`.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Implicit.Remoting.Tests.ps1:
```diff
@@ -767,7 +767,7 @@ Describe "Import-PSSession functional tests" -tags "Feature" {
 
     It "Helper functions should not be imported" -Skip:$skipTest {
-        ((Get-Item function:*PSImplicitRemoting* -ErrorAction SilentlyContinue) -eq $null) | Should Be $true
+        ($null -eq (Get-Item function:*PSImplicitRemoting* -ErrorAction SilentlyContinue)) | Should Be $true
```
Please change to `Should BeNullOrEmpty`.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Implicit.Remoting.Tests.ps1:
```diff
@@ -808,11 +808,11 @@ Describe "Import-PSSession functional tests" -tags "Feature" {
 
         It "Temporary module should be automatically removed after runspace is closed" -Skip:$skipTest {
-            ((Get-Module | Where-Object { $_.Path -eq $module.Path }) -eq $null) | Should Be $true
+            ($null -eq (Get-Module | Where-Object { $_.Path -eq $module.Path })) | Should Be $true
```
Please change to `Should BeNullOrEmpty`.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Implicit.Remoting.Tests.ps1:
```diff
@@ -808,11 +808,11 @@ Describe "Import-PSSession functional tests" -tags "Feature" {
 
         It "Temporary psm1 file should be automatically removed after runspace is closed" -Skip:$skipTest {
-            ((Get-Item $module.Path -ErrorAction SilentlyContinue) -eq $null) | Should Be $true
+            ($null -eq (Get-Item $module.Path -ErrorAction SilentlyContinue)) | Should Be $true
```
Please change to `Should BeNullOrEmpty`.

iSazonov commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Implicit.Remoting.Tests.ps1:
```diff
@@ -1737,10 +1737,10 @@ Describe "Implicit remoting tests" -tags "Feature" {
         }
 
-        ((Get-Item function:Get-MyVariable -ErrorAction SilentlyContinue) -eq $null) | Should Be $true
+        ($null -eq (Get-Item function:Get-MyVariable -ErrorAction SilentlyContinue)) | Should Be $true
```
Please change to `Should BeNullOrEmpty`.

iSazonov commented on test/powershell/engine/SemanticVersion.Tests.ps1:
```diff
@@ -117,9 +117,9 @@ Describe "SemanticVersion api tests" -Tags 'CI' {
             $null -eq $operand | Should Be $false
-            $operand -eq $null | Should Be $false
+            $null -eq $operand | Should Be $false
+            $null -ne $operand | Should Be $true
```
I believe it is special tests - please revert the changes in the file.

bergmeister committed:
Implement suggestions made by code review of PR #4261 (PSScriptAnalyzer fixes by category):
-Spacing in build.psm1 and apache-demo.ps1
-Correction of logical mistake in Implicit.Remoting.Tests.ps1
-Changing more tests from '$null -eq $object | Should Be $true' to '$object | Should BeNullOrEmpty'
-Revert a change initially made to SemanticVersion.Tests.ps1

iSazonov commented:
LGTM.

daxian-dbw commented:
I don't think we should touch CSharp files and Markdown files. PSScriptAnalyzer warnings should be applicable only to powershell script files.

iSazonov commented:
Should we fix CSharp files and Markdown files in separate PRs?

daxian-dbw commented:
I don't think we need to make those changes in non-powershell-script files, since the goal is to avoid PSScriptAnalyzer warnings.

bergmeister commented:
@daxian-dbw If you say that the only goal is to avoid PSScriptAnalyzer warnings, then I'm not sure if you considered the reason why those rules exist. They are there best practices to avoid unintended/wrong behaviour and therefore apply to any file using PS syntax IMHO. PSPossibleIncorrectComparisonWithNull is the best example where it can result in e.g. an if condition that always evaluates to true. I can understand if you are concerned that in the case of PSPossibleIncorrectComparisonWithNull, changes in the code of the actual product could lead to a regression and should be excluded but in the case of PSAvoidUsingCmdletAliases, I do not see a reason as to why cs/md files should be excluded. WDYT?

daxian-dbw commented:
@bergmeister Thanks for the comment, and you have very good points. Yes, my main concern about changing the `.cs` files is the potential regression. Given the size of changes, it could be easy for us to overlook something in the review. That being said, I think you are right that the scripts declared in C# code should follow the rules as well, but I would suggest that we leave the comments unchanged in C# code, especially those that still use other aliases.

As for the PSScriptAnalyzer rules, it's very useful to write easy-to-maintain and less error-prone PS script files, but it's not necessarily always applicable to any PSH syntax, for example, the script that serves as a one-liner command for you to run interactively, in which case aliases may be more preferred because they make the command more concise and easy to type. The changes in [`KnownIssues.md`](https://github.com/PowerShell/PowerShell/pull/4261/files#diff-f25db63262b5a4a08e655687e9e2ec18R53) would be a good example of it -- `ls (gci *.txt | % name)` is a command for a user to run interactively, and thus it's OK to keep the aliases. It's a similar idea for scripts in c# comments -- those simple scripts are only for demonstrating the scenarios, and thus it's OK to use the aliases to keep them concise.

iSazonov commented:
Only we don't type samples from code and md files - we copy-paste 😄

daxian-dbw commented:
@iSazonov right :) my point is that for the one-liner scripts like those, it's OK to use aliases to keep them concise.
I will do a pass of review on the `.cs` files.

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Implicit.Remoting.Tests.ps1:
```diff
@@ -417,7 +417,7 @@ Describe "Proxy module is usable when the original runspace is no longer around"
         It "Verifies Remove-Module removed the runspace that was automatically created" -Pending {
             Remove-Module $module -Force
-            ((Get-PSSession -InstanceId $internalSession.InstanceId -ErrorAction SilentlyContinue) -eq $null) | Should Be $true
+            (Get-PSSession -InstanceId $internalSession.InstanceId -ErrorAction SilentlyContinue) | Should BeNullOrEmpty
```
Should be `Should Be $null`

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Implicit.Remoting.Tests.ps1:
```diff
@@ -452,7 +452,7 @@ Describe "Proxy module is usable when the original runspace is no longer around"
         It "Verifies Remove-Module removes automatically created runspace" -Pending {
             Remove-Module $module -Force
-            ((Get-PSSession -InstanceId $internalSession.InstanceId -ErrorAction SilentlyContinue) -eq $null) | Should Be $true
+            (Get-PSSession -InstanceId $internalSession.InstanceId -ErrorAction SilentlyContinue) | Should BeNullOrEmpty
```
ditto

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Implicit.Remoting.Tests.ps1:
```diff
@@ -668,7 +668,7 @@ Describe "Import-PSSession with FormatAndTypes" -tags "Feature" {
         $results.Count | Should Be 2
         # First object shouldn't have the additional ETS note property
-        $results[0].MyTestLabel -eq $null | Should Be $true
+        $null -eq $results[0].MyTestLabel | Should Be $true
```
Use `Should Be $null` ?

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Implicit.Remoting.Tests.ps1:
```diff
@@ -767,7 +767,7 @@ Describe "Import-PSSession functional tests" -tags "Feature" {
 
     It "Helper functions should not be imported" -Skip:$skipTest {
-        ((Get-Item function:*PSImplicitRemoting* -ErrorAction SilentlyContinue) -eq $null) | Should Be $true
+        (Get-Item function:*PSImplicitRemoting* -ErrorAction SilentlyContinue) | Should BeNullOrEmpty
```
Ditto

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Implicit.Remoting.Tests.ps1:
```diff
@@ -800,19 +800,19 @@ Describe "Import-PSSession functional tests" -tags "Feature" {
 
         It "Temporary module should be automatically removed after runspace is closed" -Skip:$skipTest {
-            ((Get-Module | ? { $_.Path -eq $module.Path }) -eq $null) | Should Be $true
+            (Get-Module | Where-Object { $_.Path -eq $module.Path }) | Should BeNullOrEmpty
```
Ditto

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Implicit.Remoting.Tests.ps1:
```diff
@@ -800,19 +800,19 @@ Describe "Import-PSSession functional tests" -tags "Feature" {
 
         It "Temporary psm1 file should be automatically removed after runspace is closed" -Skip:$skipTest {
-            ((Get-Item $module.Path -ErrorAction SilentlyContinue) -eq $null) | Should Be $true
+            (Get-Item $module.Path -ErrorAction SilentlyContinue) | Should BeNullOrEmpty
```
Ditto

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Implicit.Remoting.Tests.ps1:
```diff
@@ -1737,10 +1737,10 @@ Describe "Implicit remoting tests" -tags "Feature" {
         }
 
-        ((Get-Item function:Get-MyVariable -ErrorAction SilentlyContinue) -eq $null) | Should Be $true
+        (Get-Item function:Get-MyVariable -ErrorAction SilentlyContinue) | Should BeNullOrEmpty
```
Ditto

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Implicit.Remoting.Tests.ps1:
```diff
@@ -1779,17 +1779,17 @@ Describe "Implicit remoting tests" -tags "Feature" {
 
                 ## Get-Variable function should not be exported when importing a BadVerb-Variable function
-                ((Get-Item Function:\Get-Variable -ErrorAction SilentlyContinue) -eq $null) | Should Be $true
+                Get-Item Function:\Get-Variable -ErrorAction SilentlyContinue | Should BeNullOrEmpty
```
Ditto

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Implicit.Remoting.Tests.ps1:
```diff
@@ -1779,17 +1779,17 @@ Describe "Implicit remoting tests" -tags "Feature" {
 
                 ## BadVerb-Variable should be a function, not an alias (1)
-                ((Get-Item Function:\BadVerb-Variable -ErrorAction SilentlyContinue) -ne $null) | Should Be $true
+                Get-Item Function:\BadVerb-Variable -ErrorAction SilentlyContinue | Should BeNullOrEmpty
```
Ditto

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Implicit.Remoting.Tests.ps1:
```diff
@@ -1779,17 +1779,17 @@ Describe "Implicit remoting tests" -tags "Feature" {
 
                 ## BadVerb-Variable should be a function, not an alias (2)
-                ((Get-Item Alias:\BadVerb-Variable -ErrorAction SilentlyContinue) -eq $null) | Should Be $true
+                Get-Item Alias:\BadVerb-Variable -ErrorAction SilentlyContinue | Should BeNullOrEmpty
```
Ditto

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Implicit.Remoting.Tests.ps1:
```diff
@@ -1819,17 +1819,17 @@ Describe "Implicit remoting tests" -tags "Feature" {
 
                 ## BadVerb-Variable should be a function, not an alias (2)
-                ((Get-Item Alias:\BadVerb-Variable -ErrorAction SilentlyContinue) -eq $null) | Should Be $true
+                Get-Item Alias:\BadVerb-Variable -ErrorAction SilentlyContinue | Should BeNullOrEmpty
```
Ditto

daxian-dbw commented on test/powershell/Modules/Microsoft.PowerShell.Utility/Implicit.Remoting.Tests.ps1:
```diff
@@ -1870,14 +1870,14 @@ Describe "Implicit remoting tests" -tags "Feature" {
 
                 ## BadVerb-Variable should be an alias, not a function (2)
-                ((Get-Item Alias:\BadVerb-Variable -ErrorAction SilentlyContinue) -ne $null) | Should Be $true
+                Get-Item Alias:\BadVerb-Variable -ErrorAction SilentlyContinue | Should BeNullOrEmpty
```
Ditto

daxian-dbw commented on test/powershell/engine/Remoting/InvokeCommandRemoteDebug.Tests.ps1:
```diff
@@ -157,12 +157,12 @@ Describe "Invoke-Command remote debugging tests" -Tags 'Feature' {
-            if ($rs2 -ne $null) { $rs2.Dispose() }
-            if ($remoteSession -ne $null) { Remove-PSSession $remoteSession -ErrorAction SilentlyContinue }
+            if ($null -ne $testDebugger) { $testDebugger.Release() }
+            if ($null -ne $psl) { $ps.Dispose() }
```
> ($null -ne $psl)

Should be `$ps`

daxian-dbw commented:
@bergmeister Please push new commits to address the comments, and please **DO NOT** rebase your branch. This way, we can keep the review history which is very important for large size changes like this PR.

daxian-dbw committed:
Merge branch 'master' into FixPSScriptAnalyzerWarningsByCategory

daxian-dbw commented:
@bergmeister I just resolved the 2 conflicts. I think we need faster review iterations on this PR, because there are many other PRs changing the script files covered in this PR and thus there will be more conflicts very soon.

bergmeister committed:
Implemented the suggested changes by the 2nd review of PR #4261 by @daxian-dbw.
This includes a revert of a logical error in Microsoft.PowerShell.ODataV4Adapter.ps1 and reverts back some functions to their alias where the reviewer felt that they were still appropriate. Some minor cosmetic changes and test syntax improvements.

bergmeister commented on src/Microsoft.PowerShell.GraphicalHost/commandHelpers/ShowCommandHelper.cs:
```diff
@@ -183,7 +183,7 @@ Function GetParameterInfo
                     [ValidateSet].IsAssignableFrom($_.GetType())
                 }
-                if (($validateSetAttributes -ne $null) -and ($validateSetAttributes.Count -gt 0))
+                if (($null -ne $validateSetAttributes) -and ($validateSetAttributes.Count -gt 0))
```
Good. Although if I had written the code myself I would have rather declared `$validateSetAttributes` as `[System.Array]` because then one would not only avoid the problem that the variable can be null but also prevents gotchas where in the case of array size 1 the type would not be an array and therefore addressing it later as $validateSetAttributes[0] would result in undesired behaviour (in most cases PowerShell would return the first character of the string representation). Just my two cents.

bergmeister commented:
@daxian-dbw I have applied your requested changes.

daxian-dbw committed:
Change 'dir' to 'Get-ChildItem'

daxian-dbw committed:
Revert the 2 changes in comments in powershell.cs

daxian-dbw committed:
Revert changes in comments in PSCommand.cs

daxian-dbw committed:
Change to 'Should Be $null' in LocalUser.Tests.ps1

daxian-dbw reviewed (approved):
LGTM. Thanks @bergmeister!

iSazonov commented:
@bergmeister Thank you for your patience and hard work!

bergmeister commented:
@iSazonov @daxian-dbw Thank you both as well. I really had no idea it would turn out to be that complex. I am currently developing a `-AutoFix` switch for `Invoke-PSScriptAnalyzer` that will make it easier for everyone to fix some of the remaining warnings in the future.
